### PR TITLE
Reduce compile time for single-device and multi-device recipes

### PR DESCRIPTION
**Pull request by yf225**
#### Context
What is the purpose of this PR? Is it to
- [ ] add a new feature
- [ ] fix a bug
- [ ] update tests and/or documentation
- [x] other (improve compile time)

Improvements in compile time (on my A100 machine):
Llama3 8B QLoRA single-device: 83s -> 47s (**-43%**)
Llama3 8B full-finetune single-device: 29s -> 20s (**-31%**)
Llama3 70B QLoRA FSDP2: 104s -> 62s (**-40%**)

#### Changelog
- No nested FSDP wrapping for LoRALinear, to reduce some recompiles.
- mark_dynamic for each layer's input `h`, to avoid dynamic sequence length induced recompiles.
- mark_dynamic for loss_fn inputs (`logits` and `labels`), to avoid dynamic sequence length induced recompiles.

#### Test plan

TODO (mentioned by Evan):
- we have CI for our recipes to make sure loss values don't change, we could add something like this to the recipe test: https://fburl.com/pusfwz1n, since we don't have it for our FSDP2 recipe yet
- How about just a stripped-down version of my test from https://github.com/pytorch/torchtune/pull/1419? Basically run for 1-2 models and show that memory and perf look good; I guess you already have compile time covered. Just paste in the memory and toks/sec into the PR desc.

Please make sure to do each of the following if applicable to your PR. (If you're not sure about any one of these just ask and we will happily help. We also have a [contributing page](https://github.com/pytorch/torchtune/blob/main/CONTRIBUTING.md) for some guidance on contributing.)

- [ ] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [ ] add [unit tests](https://github.com/pytorch/torchtune/tree/main/tests/torchtune) for any new functionality
- [ ] update [docstrings](https://github.com/pytorch/torchtune/tree/main/docs/source) for any new or updated methods or classes
- [ ] run unit tests via `pytest tests`
- [ ] run recipe tests via `pytest tests -m integration_test`
- [ ] manually run any new or modified recipes with sufficient proof of correctness
- [ ] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

#### UX
If your function changed a public API, please add a dummy example of what the user experience will look like when calling it.
Example of docstring: https://github.com/pytorch/torchtune/blob/6a7951f1cdd0b56a9746ef5935106989415f50e3/torchtune/modules/vision_transformer.py#L285
Example in our docs: https://pytorch.org/torchtune/main/tutorials/qat_finetune.html#applying-qat-to-llama3-models

- [ ] I did not change any public API;
- [ ] I have added an example to docs or docstrings;

### CODE REVIEW

File: torchtune/modules/transformer.py [hunk 1]
```diff
@@ -449,6 +449,7 @@ def forward(
             if i in self.output_hidden_states:
                 hidden.append(h)
             # shape: [b, s, d]
+            torch._dynamo.mark_dynamic(h, 1)
```
Comment (ebsmothers) [inline torchtune/modules/transformer.py:452]: Noob question on where we apply `mark_dynamic`: is it here because this is the entry point for per-layer compile? (And separately would there be any benefit to compiling the token embeddings? I guess not?)

And similarly on logits and labels: because they are the entry point to the compiled loss?

Comment (yf225) [reply torchtune/modules/transformer.py:452]: > is it here because this is the entry point for per-layer compile?

> And similarly on logits and labels: because they are the entry point to the compiled loss?

yes to both :)

> separately would there be any benefit to compiling the token embeddings? I guess not?

was thinking about it as well, and I just quickly tested it - seems no perf benefit from compiling the token embeddings :(

File: recipes/lora_finetune_single_device.py [hunk 2]
```diff
@@ -554,6 +554,8 @@ def _loss_step(self, batch: Dict[str, torch.Tensor]) -> torch.Tensor:
         labels = labels[..., 1:].contiguous()
         logits = logits.transpose(1, 2)
         # Compute loss
+        torch._dynamo.mark_dynamic(logits, 2)
+        torch._dynamo.mark_dynamic(labels, 1)
```
Comment (felipemello1) [inline recipes/lora_finetune_single_device.py:558]: does this matter if we are not compiling loss_step anymore?

**Comment by felipemello1** [comment]
@yf225 , i have just landed the chunked CE. In that PR, i also changed the shift logic:

```
# Shift so that tokens < n predict n
logits = logits[..., :-1, :].contiguous()
labels = labels[..., 1:].contiguous()
logits = logits.transpose(1, 2)
```

and added chunking to the transformer + CE loss. I would be glad to help with testing the dynamic flags with you, if you need help.

File: recipes/lora_finetune_single_device.py (hunk 2 above)
Comment (yf225) [reply recipes/lora_finetune_single_device.py:558]: yes this influence how `self._loss_fn` is compiled (currently for per-layer compile, we also compile the loss_fn)
